invenio_rdm_records/ext.py

```diff
@@ -54,14 +56,36 @@ def init_config(self, app):
                 app.config.setdefault(k, getattr(config, k))
 
 
+def prepare_publication_date(record_dict):
+    """
+    Add search compatible publication date to record_dict prior to ES index.
+
+    This date is the lowest year-month-day date from the interval or (partial)
+    date. The regular publication_date is not in a format ES can use for
+    powerful date queries.
+
+    :param record_dict: dumped Record dict
+    """
+    print("record_dict", record_dict)
+    parser = level0Expression("level0")
+    date_or_interval = parser.parseString(record_dict['publication_date'])[0]
+    # lower_strict() is available for EDTF Interval AND Date objects
+    date_tuple = date_or_interval.lower_strict()
+    record_dict['_publication_date_search'] = time.strftime(
+        "%Y-%m-%d", date_tuple
+    )
+
+
 def before_record_index_hook(
         sender, json=None, record=None, index=None, **kwargs):
     """Hook to transform Deposits before indexing in ES.
 
     :param sender: The entity sending the signal.
     :param json: The dumped Record dict which will be indexed.
-    :param record: The correspondng Record object.
+    :param record: The corresponding Record object.
     :param index: The index in which the json will be indexed.
     :param kwargs: Any other parameters.
     """
-    add_es_metadata_extensions(json)  # mutates json
+    # All thee operations mutate the json
+    add_es_metadata_extensions(json)  # TODO: Change for prepare
+    prepare_publication_date(json)
```

Comment:
Moderate: Perhaps this is easiest to discuss in a video chat. Should the "_publication_date_search" go 1) only to the index or 2) also to the record?

Reply:
Discussed via chat. We will add it to the record as well because:
- it is needed for enough serialization cases (saves us from having to edtf parse each time)
- supports our effort to keep ES record and DB record aligned  

`post_load` approach will be used for now. We will revisit when the deposit flow is hammered out.